### PR TITLE
doc: add v3 api doc

### PR DESCRIPTION
**Pull request by xiang90**
Before we start doc the API itself, let's first doc the guarantees we have.

/cc @heyitsanthony @gyuho @joshix 

@aphyr we would love to see your opinions on this doc too. Thanks for your previous input on etcdv2 doc.

### CODE REVIEW

File: Documentation/api_v3.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+- Write APIs
+    - put
+    - delete
+- Combination (read-modify-write) APIs
+    - txn
+
+### etcd Specific Definitions
+
+- operation completed
+    - An etcd operation is considered complete when it is committed through consensus, and therefore “executed” -- permanently stored -- by the etcd storage engine. The client knows an operation is completed when it receives a response from the etcd server. Note that the client may be uncertain about the status of an operation if it times out, or there is a network disruption between the client and the etcd member. etcd may also abort operations when there is a leader election. etcd does not send `abort` responses to  clients’ outstanding requests in this event.
```
Comment (gyuho) [inline Documentation/api_v3.md]: `permanently stored on disk`?

Comment (joshix) [inline Documentation/api_v3.md]: I would say we do not actually want to be that specific. Committed to permanent storage is the key idea, whatever that storage may be. 

(Curiousity: is there a lot of `fsync()`'ing or `bdflush()` kind of calls in etcd? That is, does it really concern itself with the disk or just do basic `write`s into the VFS?)

Comment (gyuho) [inline Documentation/api_v3.md]: You are right. Please ignore my comment. Thanks!

File: Documentation/api_v3.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+TODO: API doc
+
+## KV API Guarantees
+
+etcd is a consistent and durable key value store with mini-transaction(TODO: link to txn doc when we have it) support. The key value store is exposed through the KV APIs. etcd tries to ensure the strongest consistency and durability guarantees for a distributed system. This specification enumerates the KV API guarantees made by etcd.
+
+### APIs to consider
+
+- Read APIs
```
Comment (joshix) [inline Documentation/api_v3.md]: Please [use `*` to denote unordered lists in markdown](https://github.com/coreos/docs/blob/master/STYLE.md#preferred-markdown-symbols)

viz:

```
* Item
* Item2
  * subitem
* Item3
```

* Item
* Item2
  * subitem
* Item3

File: Documentation/api_v3.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    - watch
+- Write APIs
+    - put
+    - delete
+- Combination (read-modify-write) APIs
+    - txn
+
+### etcd Specific Definitions
+
+- operation completed
```
Comment (joshix) [inline Documentation/api_v3.md]: `#### Operation completed`

File: Documentation/api_v3.md (hunk 1 above)
Comment (joshix) [inline Documentation/api_v3.md]: remove - and make a regular paragraph below the new "Operation completed" heading

File: Documentation/api_v3.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    - An etcd operation is considered complete when it is committed through consensus, and therefore “executed” -- permanently stored -- by the etcd storage engine. The client knows an operation is completed when it receives a response from the etcd server. Note that the client may be uncertain about the status of an operation if it times out, or there is a network disruption between the client and the etcd member. etcd may also abort operations when there is a leader election. etcd does not send `abort` responses to  clients’ outstanding requests in this event.
+
+- revision
```
Comment (joshix) [inline Documentation/api_v3.md]: `#### Revision` (as above)

File: Documentation/api_v3.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    - An etcd operation that modifies the key value store is assigned with a single increasing revision. A transaction operation might modifies the key value store multiple times, but only one revision is assigned. The revision attribute of a key value pair that modified by the operation has the same value as the revision of the operation. The revision can be used as a logical clock for key value store. A key value pair that has a larger revision is modified after a key value pair with a smaller revision. Two key value pairs that have the same revision are modified by an operation "concurrently".
```
Comment (joshix) [inline Documentation/api_v3.md]: make a paragraph not a list item. (as above)

File: Documentation/api_v3.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+#### Consistency
+
+All API calls ensure [sequential consistency][seq_consistency], the strongest consistency guarantee available from distributed systems. No matter which etcd member server a client makes requests to, a client reads the same events in the same order. If two members complete the same number of operations, the state of the two members is consistent.
+
+For watch operations, etcd guarantees to return the same value for the same key across all members for the same revision. For range operations, etcd has a similar guarantee for [linearized][Linearizability] access; serialized access may be behind the quorum state, so that the later revision is not yet available.
+
+As with all distributed systems, it is impossible for etcd to ensure [strict consistency][strict_consistency]. etcd does not guarantee that it will return to a read the “most recent” value (as measured by a wall clock when request is returned) available on any cluster member.
```
Comment (joshix) [inline Documentation/api_v3.md]: "...when **a** request is **received**..." (a request isn't returned; a reply or response is returned. a request is received.)

File: Documentation/api_v3.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+Any completed operations are durable. All accessible data is also durable data. A read will never return data that has not been made durable.
+
+#### Linearizability
+
+Linearizability (also known as Atomic Consistency or External Consistency) is a consistency level between strict consistency and sequential consistency. 
+
+For linearizability, suppose each operation receives a timestamp from a loosely synchronized global clock. Operations are linearized if and only if they always complete as though they were executed in a sequential order and each operation appears to complete in the order specified by the program. Likewise, if an operation’s timestamp precedes another, that operation must also precede the other operation in the sequence.
+
+For example, if a client completes a write at `T1`. A client issues a read at `T2` (`T2` > `T1`) should read the value as least as new as the previous write completed at `T1`. But note that the read might actually completes at `T3` and the the returned value might be stale at `T3`.
```
Comment (joshix) [inline Documentation/api_v3.md]: For example, consider a client completing a write at time point 1 (*t1*). A client issuing a read at *t2* (for *t2* > *t1*) should receive a value at least as recent as the previous write, completed at *t1*. However, the read might actually complete only by *t3*, and the returned value, current at *t2* when the read began, might be "stale" by *t3*.

File: Documentation/api_v3.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+#### Linearizability
+
+Linearizability (also known as Atomic Consistency or External Consistency) is a consistency level between strict consistency and sequential consistency. 
+
+For linearizability, suppose each operation receives a timestamp from a loosely synchronized global clock. Operations are linearized if and only if they always complete as though they were executed in a sequential order and each operation appears to complete in the order specified by the program. Likewise, if an operation’s timestamp precedes another, that operation must also precede the other operation in the sequence.
+
+For example, if a client completes a write at `T1`. A client issues a read at `T2` (`T2` > `T1`) should read the value as least as new as the previous write completed at `T1`. But note that the read might actually completes at `T3` and the the returned value might be stale at `T3`. 
+
+etcd does not ensure linearizability for watch operations. We expect users to verify the revision of watch responses to ensure correct ordering.
```
Comment (joshix) [inline Documentation/api_v3.md]: Users are expected to verify the revision of watch responses to ensure correct ordering.

File: Documentation/api_v3.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+etcd ensures linearizability for all other operations by default. Linearizability comes with a cost, however, because linearized requests must go through the Raft consensus process. To obtain lower latencies and higher throughput for read requests, clients can configure a request’s consistency mode to `serializable`, which may access stale data with respect to quorum, but removes the performance penalty of linearized accesses reliance on live consensus.
```
Comment (joshix) [inline Documentation/api_v3.md]: linearized accesses' (plural possessive apostrophe)

Comment (aphyr) [inline Documentation/api_v3.md]: unsubscribe

**Comment by xiang90** [comment]
@joshix All fixed. PTAL.

**Comment by joshix** [comment]
LGTM

**Pull request merged**